runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyPipelineFuser.java

```diff
@@ -69,6 +69,15 @@ private GreedyPipelineFuser(Pipeline p) {
     fusePipeline(groupSiblings(rootConsumers));
   }
 
+  /**
+   * Fuses a {@link Pipeline} into a collection of {@link ExecutableStage}s.
+   *
+   * <p>This fuser expects each ExecutableStage to have exactly one input. This means that pipelines
+   * must be rooted at Impulse, or other runner-executed primitive transforms, instead of primitive
+   * Read nodes. The utilities in
+   * {@link org.apache.beam.runners.core.construction.JavaReadViaImpulse} can be used to translate
+   * non-compliant pipelines.
```

Comment:
This does kind of have an associated `TODO` for unbounded reads; https://issues.apache.org/jira/browse/BEAM-3859 is the (just-authored) issue to link against.

'can be used to translate non-compliant pipelines -> can be used to convert bounded pipelines using the `Read` primitive.'

Reply:
Done.